File: impl/go/compile_candidates.go
package twill

import (
	"math/big"
	"sort"
)

// CompileFlags control candidate compilation.
type CompileFlags int

const (
	// RespectImportant applies the design-system-wide important flag.
	RespectImportant CompileFlags = 1 << 0
)

// PropertySort is the sort key of a node list (SPEC §11.3).
type PropertySort struct {
	Order []int
	Count int
}

// CompiledRule is a compiled rule with its property sort key.
type CompiledRule struct {
	Node         *Rule
	PropertySort PropertySort
}

// GetPropertySort computes the property sort key of a node list.
func GetPropertySort(nodes []Node) PropertySort {
	seen := map[int]bool{}
	count := 0
	queue := append([]Node{}, nodes...)
outer:
	for len(queue) > 0 {
		n := queue[0]
		queue = queue[1:]
		switch n := n.(type) {
		case *Declaration:
			if n.NoValue {
				continue
			}
			count++
			if n.Property == "--tw-sort" {
				if index := PropertyIndex(n.Value); index != -1 {
					seen[index] = true
					break outer
				}
				continue
			}
			if index := PropertyIndex(n.Property); index != -1 {
				seen[index] = true
			}
		case *Rule:
			queue = append(queue, n.Nodes...)
		case *AtRule:
			queue = append(queue, n.Nodes...)
		case *Context:
			queue = append(queue, n.Nodes...)
		}
	}
	order := make([]int, 0, len(seen))
	for index := range seen {
		order = append(order, index)
	}
	sort.Ints(order)
	return PropertySort{Order: order, Count: count}
}

func compileBaseUtility(c *Candidate, ds *DesignSystem) [][]Node {
	if c.Kind == CandidateArbitrary {
		value, ok := AsColor(c.ArbitraryValue, c.Modifier, ds.Theme)
		if !ok {
			return nil
		}
		return [][]Node{{Decl(c.Property, value)}}
	}
	var defs []*UtilityDefinition
	for _, d := range ds.Utilities.Get(c.Root) {
		if (d.Kind == UtilityStatic) == (c.Kind == CandidateStatic) {
			defs = append(defs, d)
		}
	}
	var ordered []*UtilityDefinition
	for _, d := range defs {
		if !d.IsFallback() {
			ordered = append(ordered, d)
		}
	}
	for _, d := range defs {
		if d.IsFallback() {
			ordered = append(ordered, d)
		}
	}
	var results [][]Node
	for _, d := range ordered {
		nodes, status := d.Compile(c)
		switch status {
		case NotHandled:
			continue
		case Invalid:
			if d.Types != nil {
				return results
			}
			continue
		}
		results = append(results, nodes)
	}
	return results
}

func applyImportant(nodes []Node) {
	Walk(&nodes, func(n Node, _ *WalkUtils) WalkAction {
		switch n := n.(type) {
		case *AtRoot:
			return Skip
		case *Declaration:
			n.Important = true
		}
		return Continue
	})
}

// compileAstNodes compiles one candidate interpretation (SPEC §11.2).
func compileAstNodes(c *Candidate, flags CompileFlags, ds *DesignSystem) []CompiledRule {
	nodeLists := compileBaseUtility(c, ds)
	if len(nodeLists) == 0 {
		return nil
	}
	important := c.Important || (ds.Important && flags&RespectImportant != 0)
	var results []CompiledRule
	for _, nodes := range nodeLists {
		propertySort := GetPropertySort(nodes)
		if important {
			applyImportant(nodes)
		}
		rule := StyleRule("."+Escape(c.Raw), nodes...)
		for _, v := range c.Variants {
			if !ApplyVariant(rule, v, ds.Variants, 0) {
				return nil
			}
		}
		wrapped := []Node{rule}
		if _, err := SubstituteFunctions(&wrapped, ds); err != nil {
			return nil
		}
		if _, err := SubstituteAtVariant(&wrapped, ds); err != nil {
			return nil
		}
		results = append(results, CompiledRule{Node: rule, PropertySort: propertySort})
	}
	return results
}

// CompileAstNodes compiles a candidate interpretation, memoized per
// candidate object and flags. Cached results are cloned.
func (ds *DesignSystem) CompileAstNodes(c *Candidate, flags CompileFlags) []CompiledRule {
	byFlags, ok := ds.compileCache[c]
	if !ok {
		byFlags = map[int][]CompiledRule{}
		ds.compileCache[c] = byFlags
	}
	cached, ok := byFlags[int(flags)]
	if !ok {
		cached = compileAstNodes(c, flags, ds)
		byFlags[int(flags)] = cached
	}
	out := make([]CompiledRule, len(cached))
	for i, r := range cached {
		out[i] = CompiledRule{Node: CloneNode(r.Node).(*Rule), PropertySort: r.PropertySort}
	}
	return out
}

// CompileCandidatesOptions configure CompileCandidates.
type CompileCandidatesOptions struct {
	// IgnoreImportant disables the design-system-wide important flag.
	IgnoreImportant bool
	// OnInvalidCandidate receives each invalid raw candidate.
	OnInvalidCandidate func(candidate string)
}

type sortedRule struct {
	node         *Rule
	propertySort PropertySort
	variantOrder *big.Int
	candidate    string
}

// CompileCandidates compiles and sorts raw candidates (SPEC §11.1, §15.3).
func CompileCandidates(rawCandidates []string, ds *DesignSystem, options CompileCandidatesOptions) []*Rule {
	onInvalid := options.OnInvalidCandidate
	if onInvalid == nil {
		onInvalid = func(string) {}
	}
	flags := RespectImportant
	if options.IgnoreImportant {
		flags = 0
	}

	type match struct {
		raw        string
		candidates []*Candidate
	}
	var matches []match
	for _, raw := range rawCandidates {
		if ds.InvalidCandidates[raw] {
			onInvalid(raw)
			continue
		}
		parsed := ds.ParseCandidate(raw)
		if len(parsed) == 0 {
			onInvalid(raw)
			continue
		}
		matches = append(matches, match{raw, parsed})
	}

	order := ds.VariantOrder()
	var rules []sortedRule
	for _, m := range matches {
		found := false
		for _, c := range m.candidates {
			for _, compiled := range ds.CompileAstNodes(c, flags) {
				found = true
				variantOrder := new(big.Int)
				for _, v := range c.Variants {
					variantOrder.SetBit(variantOrder, order[v], 1)
				}
				rules = append(rules, sortedRule{compiled.Node, compiled.PropertySort, variantOrder, m.raw})
			}
		}
		if !found {
			onInvalid(m.raw)
		}
	}

	sort.SliceStable(rules, func(i, j int) bool { return compareSortedRules(rules[i], rules[j]) < 0 })
	out := make([]*Rule, len(rules))
	for i, r := range rules {
		out[i] = r.node
	}
	return out
}

func compareSortedRules(a, z sortedRule) int {
	if c := a.variantOrder.Cmp(z.variantOrder); c != 0 {
		return c
	}
	length := len(a.propertySort.Order)
	if len(z.propertySort.Order) > length {
		length = len(z.propertySort.Order)
	}
	for i := 0; i < length; i++ {
		ai, zi := -1, -1
		if i < len(a.propertySort.Order) {
			ai = a.propertySort.Order[i]
		}
		if i < len(z.propertySort.Order) {
			zi = z.propertySort.Order[i]
		}
		if ai == zi {
			continue
		}
		// A missing entry counts as infinity.
		if ai == -1 {
			return 1
		}
		if zi == -1 {
			return -1
		}
		return ai - zi
	}
	if a.propertySort.Count != z.propertySort.Count {
		return z.propertySort.Count - a.propertySort.Count
	}
	return Compare(a.candidate, z.candidate)
}

File: impl/ocaml/src/compile_candidates.ml
(* Candidate compilation and ordering (SPEC §11). *)

open Ast
open Candidate
open Design_system

(* Applies the design-system-wide important flag. *)
let respect_important = 1

(* Computes the property sort key of a node list. *)
let get_property_sort nodes =
  let seen = Hashtbl.create 16 and count = ref 0 in
  let queue = Queue.create () in
  List.iter (fun n -> Queue.add n queue) nodes;
  (try
     while not (Queue.is_empty queue) do
       match Queue.pop queue with
       | Declaration d when d.no_value -> ()
       | Declaration d ->
           incr count;
           if d.property = "--tw-sort" then begin
             let index = Property_order.index d.value in
             if index <> -1 then begin
               Hashtbl.replace seen index ();
               raise Exit
             end
           end
           else
             let index = Property_order.index d.property in
             if index <> -1 then Hashtbl.replace seen index ()
       | Rule r -> List.iter (fun n -> Queue.add n queue) !(r.rule_nodes)
       | At_rule a -> List.iter (fun n -> Queue.add n queue) !(a.at_nodes)
       | Context c -> List.iter (fun n -> Queue.add n queue) !(c.ctx_nodes)
       | _ -> ()
     done
   with Exit -> ());
  let order = List.sort compare (Hashtbl.fold (fun k () acc -> k :: acc) seen []) in
  { order; count = !count }

let compile_base_utility (c : candidate) (ds : Design_system.t) =
  if c.ckind = Arbitrary_candidate then
    match Utilities.as_color c.arbitrary_value c.cmodifier ds.theme with
    | Some value -> [ [ decl c.property value ] ]
    | None -> []
  else begin
    let defs =
      List.filter
        (fun (d : Utilities.definition) -> d.ukind = Static = (c.ckind = Static_candidate))
        (Utilities.get ds.utilities c.croot)
    in
    let ordered = List.filter (fun d -> not (Utilities.is_fallback d)) defs @ List.filter Utilities.is_fallback defs in
    let results = ref [] in
    (try
       List.iter
         (fun (d : Utilities.definition) ->
           match d.compile c with
           | _, Utilities.Not_handled -> ()
           | _, Utilities.Invalid -> if d.types <> None then raise Exit
           | nodes, Utilities.Handled -> results := nodes :: !results)
         ordered
     with Exit -> ());
    List.rev !results
  end

let apply_important nodes =
  ignore
    (walk (ref nodes) (fun n _ ->
         match n with
         | At_root _ -> Skip
         | Declaration d ->
             d.important <- true;
             Continue
         | _ -> Continue))

(* Compiles one candidate interpretation (SPEC §11.2). *)
let compile_ast_nodes (c : candidate) flags (ds : Design_system.t) =
  let node_lists = compile_base_utility c ds in
  if node_lists = [] then []
  else begin
    let important = c.important || (ds.important && flags land respect_important <> 0) in
    try
      List.map
        (fun nodes ->
          let property_sort = get_property_sort nodes in
          if important then apply_important nodes;
          let r = { selector = "." ^ Utils.escape c.raw; rule_nodes = ref nodes } in
          if not (List.for_all (fun v -> Variants.apply_variant (Rule r) v ds.variants 0) c.variants) then raise Exit;
          let wrapped = ref [ Rule r ] in
          (try
             ignore (Theme_functions.substitute_functions wrapped ds);
             ignore (At_variant.substitute wrapped ds)
           with Twill_error.Error _ -> raise Exit);
          { node = r; property_sort })
        node_lists
    with Exit -> []
  end

(* Compiles a candidate interpretation, memoized per candidate and flags.
   Cached results are cloned. *)
let compile_ast_nodes_cached (ds : Design_system.t) (c : candidate) flags =
  let cached =
    match Hashtbl.find_opt ds.compile_cache (c, flags) with
    | Some cached -> cached
    | None ->
        let compiled = compile_ast_nodes c flags ds in
        Hashtbl.replace ds.compile_cache (c, flags) compiled;
        compiled
  in
  List.map
    (fun r ->
      match clone_node (Rule r.node) with Rule cloned -> { r with node = cloned } | _ -> assert false)
    cached

type sorted_rule = { srule : Ast.rule; sort : property_sort; variant_order : int list; candidate : string }

let compare_sorted_rules a z =
  let c = compare a.variant_order z.variant_order in
  if c <> 0 then c
  else
    let rec go ao zo =
      match (ao, zo) with
      | [], [] -> 0
      | [], _ -> 1
      | _, [] -> -1
      | ai :: ar, zi :: zr -> if ai = zi then go ar zr else ai - zi
    in
    let c = go a.sort.order z.sort.order in
    if c <> 0 then c
    else if a.sort.count <> z.sort.count then z.sort.count - a.sort.count
    else Utils.compare_natural a.candidate z.candidate

(* Compiles and sorts raw candidates (SPEC §11.1). [on_invalid] receives
   each invalid raw candidate. *)
let compile_candidates ?(ignore_important = false) ?(on_invalid = fun _ -> ()) raw_candidates (ds : Design_system.t) =
  let flags = if ignore_important then 0 else respect_important in
  let matches =
    List.filter_map
      (fun raw ->
        if Hashtbl.mem ds.invalid_candidates raw then (on_invalid raw; None)
        else
          match Design_system.parse_candidate ds raw with
          | [] -> on_invalid raw; None
          | parsed -> Some (raw, parsed))
      raw_candidates
  in
  let order = Design_system.variant_order ds in
  let rules =
    List.concat_map
      (fun (raw, candidates) ->
        let found = ref false in
        let rules =
          List.concat_map
            (fun (c : candidate) ->
              List.map
                (fun (compiled : compiled_rule) ->
                  found := true;
                  let bits =
                    List.sort_uniq (fun a b -> compare b a)
                      (List.map (fun (v : variant) -> match Hashtbl.find_opt order v.vraw with Some i -> i | None -> 0) c.variants)
                  in
                  { srule = compiled.node; sort = compiled.property_sort; variant_order = bits; candidate = raw })
                (compile_ast_nodes_cached ds c flags))
            candidates
        in
        if not !found then on_invalid raw;
        rules)
      matches
  in
  List.map (fun r -> r.srule) (List.stable_sort compare_sorted_rules rules)
